File: android/core-domain/src/main/kotlin/com/crispy/tv/domain/media/MediaStateContract.kt
package com.crispy.tv.domain.media

import com.crispy.tv.domain.MediaKey

data class MediaStateNormalized(
    val cardFamily: String? = null,
    val mediaKey: MediaKey? = null,
    val mediaType: String? = null,
    val itemId: String? = null,
    val title: String? = null,
    val posterUrl: String? = null,
    val backdropUrl: String? = null,
    val subtitle: String? = null,
    val progressPercent: Double? = null,
    val watchedAt: String? = null,
    val lastActivityAt: String? = null,
    val origins: List<String>? = null,
    val dismissible: Boolean? = null,
    val layout: String? = null,
    val routeKind: String? = null,
)

fun normalizeMediaStateCard(payload: Map<String, Any?>, kind: String): MediaStateNormalized? {
    return when (kind.trim().lowercase()) {
        "media_item", "search_result", "recommendation_item" -> normalizeBaseItemDto(payload)
        "continue_watching_item" -> normalizeContinueWatching(payload)
        "watched_item" -> normalizeWatchedItem(payload)
        "watchlist_item" -> normalizeWatchlistItem(payload)
        "rating_item" -> normalizeRatingItem(payload)
        "library_item" -> normalizeLibraryItem(payload)
        "home_snapshot_section" -> normalizeHomeSnapshotSection(payload)
        "title_route" -> normalizeTitleRoute(payload)
        else -> null
    }
}

private fun normalizeBaseItemDto(payload: Map<String, Any?>): MediaStateNormalized? {
    val mediaKeyStr = payload.stringValue("Id") ?: return null
    val mediaType = payload.stringValue("Type") ?: return null
    val title = payload.stringValue("Name") ?: return null
    val imageTags = payload.objectValue("ImageTags")
    return MediaStateNormalized(
        cardFamily = "media_item",
        mediaKey = MediaKey(mediaKeyStr),
        mediaType = mediaType,
        title = title,
        posterUrl = imageTags?.imageTagMedium("Primary"),
        backdropUrl = imageTags?.backdropMedium(),
        subtitle = payload.nullableStringValue("EpisodeTitle") ?: payload.nullableStringValue("Overview"),
    )
}

private fun userDataFrom(payload: Map<String, Any?>): Map<String, Any?>? {
    return payload.objectValue("UserData")
}

private fun normalizeContinueWatching(payload: Map<String, Any?>): MediaStateNormalized? {
    val normalized = normalizeBaseItemDto(payload) ?: return null
    val ud = userDataFrom(payload)
    return normalized.copy(
        itemId = normalized.mediaKey?.value,
        progressPercent = ud?.doubleValue("PlayedPercentage") ?: 0.0,
        lastActivityAt = ud?.stringValue("LastPlayedDate") ?: payload.stringValue("LastPlayedDate"),
        dismissible = ud?.booleanValue("DismissedFromContinueWatching") ?: false,
        origins = emptyList(),
    )
}

private fun normalizeWatchedItem(payload: Map<String, Any?>): MediaStateNormalized? {
    val normalized = normalizeBaseItemDto(payload) ?: return null
    val ud = userDataFrom(payload)
    return normalized.copy(
        itemId = normalized.mediaKey?.value,
        watchedAt = ud?.stringValue("LastPlayedDate"),
        origins = emptyList(),
    )
}

private fun normalizeWatchlistItem(payload: Map<String, Any?>): MediaStateNormalized? {
    val normalized = normalizeBaseItemDto(payload) ?: return null
    return normalized.copy(
        itemId = normalized.mediaKey?.value,
        origins = emptyList(),
    )
}

private fun normalizeRatingItem(payload: Map<String, Any?>): MediaStateNormalized? {
    val normalized = normalizeBaseItemDto(payload) ?: return null
    return normalized.copy(
        itemId = normalized.mediaKey?.value,
        origins = emptyList(),
    )
}

private fun normalizeLibraryItem(payload: Map<String, Any?>): MediaStateNormalized? {
    val normalized = normalizeBaseItemDto(payload) ?: return null
    return normalized.copy(
        itemId = normalized.mediaKey?.value,
        origins = emptyList(),
    )
}

private fun normalizeHomeSnapshotSection(payload: Map<String, Any?>): MediaStateNormalized? {
    val layout = payload.stringValue("layout") ?: return null
    if (layout !in setOf("regular", "landscape", "collection", "hero")) return null
    val items = payload["items"] as? List<*> ?: return null
    if (items.isEmpty()) return null
    return MediaStateNormalized(layout = layout)
}

private fun normalizeTitleRoute(payload: Map<String, Any?>): MediaStateNormalized? {
    val mediaKeyStr = payload.stringValue("mediaKey") ?: payload.stringValue("Id") ?: return null
    val path = payload.stringValue("path") ?: return null
    if (path != "/v1/metadata/titles/$mediaKeyStr") return null
    return MediaStateNormalized(mediaKey = MediaKey(mediaKeyStr), routeKind = "title")
}

private fun Map<String, Any?>.stringValue(key: String): String? {
    return this[key]?.toString()?.trim()?.takeIf { it.isNotEmpty() }
}

private fun Map<String, Any?>.nullableStringValue(key: String): String? {
    if (!containsKey(key)) return null
    return this[key]?.toString()?.trim()?.takeIf { it.isNotEmpty() }
}

private fun Map<String, Any?>.objectValue(key: String): Map<String, Any?>? {
    @Suppress("UNCHECKED_CAST")
    return this[key] as? Map<String, Any?>
}

private fun Map<String, Any?>.imageTagMedium(key: String): String? {
    val tag = this[key]
    return when (tag) {
        is String -> tag.trim().ifBlank { null }
        is Map<*, *> -> tag.mapStringValue("medium")
        else -> null
    }
}

private fun Map<String, Any?>.backdropMedium(): String? {
    val backdrops = this["Backdrop"] as? List<*> ?: return null
    val first = backdrops.firstOrNull() ?: return null
    return when (first) {
        is String -> first.trim().ifBlank { null }
        is Map<*, *> -> first.mapStringValue("medium")
        else -> null
    }
}

private fun Map<*, *>.mapStringValue(key: String): String? {
    return this[key]?.toString()?.trim()?.takeIf { it.isNotEmpty() }
}

private fun Map<String, Any?>.booleanValue(key: String): Boolean? {
    return when (val value = this[key]) {
        is Boolean -> value
        is String -> when (value.trim().lowercase()) {
            "true" -> true
            "false" -> false
            else -> null
        }
        else -> null
    }
}

private fun Map<String, Any?>.doubleValue(key: String): Double? {
    return when (val value = this[key]) {
        is Number -> value.toDouble()
        is String -> value.trim().toDoubleOrNull()
        else -> null
    }
}

File: ios/ContractRunner/Sources/ContractRunner/MediaStateContract.swift
import Foundation

public struct MediaStateNormalized: Equatable {
    public let cardFamily: String?
    public let itemId: String?
    public let mediaType: String?
    public let title: String?
    public let posterUrl: String?
    public let backdropUrl: String?
    public let subtitle: String?
    public let progressPercent: Double?
    public let watchedAt: String?
    public let lastActivityAt: String?
    public let origins: [String]?
    public let dismissible: Bool?
    public let layout: String?
    public let routeKind: String?

    public init(
        cardFamily: String? = nil,
        itemId: String? = nil,
        mediaType: String? = nil,
        title: String? = nil,
        posterUrl: String? = nil,
        backdropUrl: String? = nil,
        subtitle: String? = nil,
        progressPercent: Double? = nil,
        watchedAt: String? = nil,
        lastActivityAt: String? = nil,
        origins: [String]? = nil,
        dismissible: Bool? = nil,
        layout: String? = nil,
        routeKind: String? = nil
    ) {
        self.cardFamily = cardFamily
        self.itemId = itemId
        self.mediaType = mediaType
        self.title = title
        self.posterUrl = posterUrl
        self.backdropUrl = backdropUrl
        self.subtitle = subtitle
        self.progressPercent = progressPercent
        self.watchedAt = watchedAt
        self.lastActivityAt = lastActivityAt
        self.origins = origins
        self.dismissible = dismissible
        self.layout = layout
        self.routeKind = routeKind
    }
}

public func normalizeMediaStateCard(payload: [String: Any], kind: String) -> MediaStateNormalized? {
    switch kind.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() {
    case "media_item", "search_result", "recommendation_item":
        return normalizeBaseItemDto(payload)
    case "continue_watching_item":
        return normalizeContinueWatching(payload)
    case "watched_item":
        return normalizeWatchedItem(payload)
    case "watchlist_item":
        return normalizeWatchlistItem(payload)
    case "rating_item":
        return normalizeRatingItem(payload)
    case "library_item":
        return normalizeLibraryItem(payload)
    case "home_snapshot_section":
        return normalizeHomeSnapshotSection(payload)
    case "title_route":
        return normalizeTitleRoute(payload)
    default:
        return nil
    }
}

private func normalizeBaseItemDto(_ payload: [String: Any]) -> MediaStateNormalized? {
    guard let itemId = stringValue(payload, "Id"),
          let mediaType = stringValue(payload, "Type"),
          let title = stringValue(payload, "Name") else {
        return nil
    }
    let imageTags = objectValue(payload, "ImageTags")
    return MediaStateNormalized(
        cardFamily: "media_item",
        itemId: itemId,
        mediaType: mediaType,
        title: title,
        posterUrl: imageTagMedium(imageTags, "Primary"),
        backdropUrl: backdropMedium(imageTags),
        subtitle: nullableStringValue(payload, "EpisodeTitle") ?? nullableStringValue(payload, "Overview")
    )
}

private func userDataFrom(_ payload: [String: Any]) -> [String: Any]? {
    return objectValue(payload, "UserData")
}

private func normalizeContinueWatching(_ payload: [String: Any]) -> MediaStateNormalized? {
    guard let normalized = normalizeBaseItemDto(payload) else { return nil }
    let ud = userDataFrom(payload)
    return MediaStateNormalized(
        cardFamily: normalized.cardFamily,
        itemId: normalized.itemId,
        mediaType: normalized.mediaType,
        title: normalized.title,
        posterUrl: normalized.posterUrl,
        backdropUrl: normalized.backdropUrl,
        subtitle: normalized.subtitle,
        progressPercent: doubleValue(ud, "PlayedPercentage") ?? 0,
        lastActivityAt: stringValue(ud, "LastPlayedDate"),
        origins: [],
        dismissible: ud?["DismissedFromContinueWatching"] as? Bool ?? false
    )
}

private func normalizeWatchedItem(_ payload: [String: Any]) -> MediaStateNormalized? {
    guard let normalized = normalizeBaseItemDto(payload) else { return nil }
    let ud = userDataFrom(payload)
    return MediaStateNormalized(
        cardFamily: normalized.cardFamily,
        itemId: normalized.itemId,
        mediaType: normalized.mediaType,
        title: normalized.title,
        posterUrl: normalized.posterUrl,
        backdropUrl: normalized.backdropUrl,
        subtitle: normalized.subtitle,
        watchedAt: stringValue(ud, "LastPlayedDate"),
        origins: []
    )
}

private func normalizeWatchlistItem(_ payload: [String: Any]) -> MediaStateNormalized? {
    guard let normalized = normalizeBaseItemDto(payload) else { return nil }
    return MediaStateNormalized(
        cardFamily: normalized.cardFamily,
        itemId: normalized.itemId,
        mediaType: normalized.mediaType,
        title: normalized.title,
        posterUrl: normalized.posterUrl,
        backdropUrl: normalized.backdropUrl,
        subtitle: normalized.subtitle,
        origins: []
    )
}

private func normalizeRatingItem(_ payload: [String: Any]) -> MediaStateNormalized? {
    guard let normalized = normalizeBaseItemDto(payload) else { return nil }
    return MediaStateNormalized(
        cardFamily: normalized.cardFamily,
        itemId: normalized.itemId,
        mediaType: normalized.mediaType,
        title: normalized.title,
        posterUrl: normalized.posterUrl,
        backdropUrl: normalized.backdropUrl,
        subtitle: normalized.subtitle,
        origins: []
    )
}

private func normalizeLibraryItem(_ payload: [String: Any]) -> MediaStateNormalized? {
    guard let normalized = normalizeBaseItemDto(payload) else { return nil }
    return MediaStateNormalized(
        cardFamily: normalized.cardFamily,
        itemId: normalized.itemId,
        mediaType: normalized.mediaType,
        title: normalized.title,
        posterUrl: normalized.posterUrl,
        backdropUrl: normalized.backdropUrl,
        subtitle: normalized.subtitle,
        origins: []
    )
}

private func normalizeHomeSnapshotSection(_ payload: [String: Any]) -> MediaStateNormalized? {
    guard let layout = stringValue(payload, "layout"),
          ["regular", "landscape", "collection", "hero"].contains(layout),
          let items = payload["items"] as? [Any], !items.isEmpty else {
        return nil
    }
    return MediaStateNormalized(layout: layout)
}

private func normalizeTitleRoute(_ payload: [String: Any]) -> MediaStateNormalized? {
    guard let itemId = stringValue(payload, "Id"),
          let path = stringValue(payload, "path"),
          path == "/v1/metadata/items/\(itemId)" else {
        return nil
    }
    return MediaStateNormalized(itemId: itemId, routeKind: "title")
}

private func stringValue(_ object: [String: Any]?, _ key: String) -> String? {
    guard let value = object?[key] as? String else {
        return nil
    }
    let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
    return trimmed.isEmpty ? nil : trimmed
}

private func nullableStringValue(_ object: [String: Any], _ key: String) -> String? {
    guard let value = object[key] else {
        return nil
    }
    if value is NSNull {
        return nil
    }
    guard let text = value as? String else {
        return nil
    }
    let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
    return trimmed.isEmpty ? nil : trimmed
}

private func doubleValue(_ object: [String: Any]?, _ key: String) -> Double? {
    guard let value = object?[key] else {
        return nil
    }
    if let double = value as? Double {
        return double
    }
    if let number = value as? NSNumber {
        return number.doubleValue
    }
    if let string = value as? String {
        return Double(string.trimmingCharacters(in: .whitespacesAndNewlines))
    }
    return nil
}

private func objectValue(_ object: [String: Any]?, _ key: String) -> [String: Any]? {
    return object?[key] as? [String: Any]
}

private func imageTagMedium(_ tags: [String: Any]?, _ key: String) -> String? {
    guard let tag = tags?[key] else { return nil }
    if let string = tag as? String { return string }
    if let dict = tag as? [String: Any] { return nullableStringValue(dict, "medium") }
    return nil
}

private func backdropMedium(_ tags: [String: Any]?) -> String? {
    guard let backdrops = tags?["Backdrop"] as? [Any], let first = backdrops.first else { return nil }
    if let string = first as? String { return string }
    if let dict = first as? [String: Any] { return nullableStringValue(dict, "medium") }
    return nil
}
